src/charm.py

```diff
@@ -87,6 +92,115 @@ def _spring_boot_port(self) -> int:
             return port
         return 8080
 
+    def _exec(
+        self, command: typing.List[str], environment: typing.Optional[typing.Dict[str, str]] = None
+    ) -> typing.Tuple[int, str, str]:
+        """Execute a command in Spring Boot application container with a timeout of 60s.
+
+        Args:
+            command: command to be executed.
+            environment: environment variables for the process.
+
+        Returns:
+            A 3-tuple of exit code, stdout and stderr.
+        """
+        container = self._spring_boot_container()
+        process = container.exec(command, environment=environment, timeout=60)
+        try:
+            stdout, stderr = process.wait_output()
+            return 0, stdout, stderr
+        except ExecError as error:
+            return error.exit_code, typing.cast(str, error.stdout), typing.cast(str, error.stderr)
+
+    def _parse_human_readable_units(self, number: str) -> int:
+        """Parse numbers with human-readable units like K, M and G.
+
+        Args:
+            number: input string, something like ``"1G"`` or ``"33m"``.
+
+        Returns:
+            Parsed result, as an integer.
+
+        Raises:
+            ValueError: when the input number is invalid.
+        """
+        number = number.lower()
+        unit = number[-1].lower()
+        unit_scale = {"k": 2**10, "m": 2**20, "g": 2**30}
+        if unit not in unit_scale:
+            try:
+                return int(number)
+            except ValueError as exc:
+                raise ValueError(f"Unknown human-readable unit: {repr(number)}") from exc
+        digits = number[:-1]
+        return int(digits) * unit_scale[unit]
+
+    def _extract_jvm_heap_initial_memory(self, jvm_config: str) -> int:
+        """Get the JVM heap initial memory size from a JVM configuration string.
+
+        Args:
+            jvm_config: JVM configuration string.
+
+        Returns:
+            JVM heap initial memory size in number of bytes. ``0`` if not defined in the JVM
+            configuration string.
+        """
+        candidates = re.findall("(^|\\s)(-Xms\\d+[kmgtKMGT]?)\\b", jvm_config)
+        if not candidates:
+            return 0
+        return self._parse_human_readable_units(candidates[-1][-1].removeprefix("-Xms"))
+
+    def _extract_jvm_heap_maximum_memory(self, jvm_config: str) -> int:
+        """Get the JVM heap maximum memory size from a JVM configuration string.
+
+        Args:
+            jvm_config: JVM configuration string.
+
+        Returns:
+            JVM heap maximum memory size in number of bytes. ``0`` if not defined in the JVM
+            configuration string.
+        """
+        candidates = re.findall("(^|\\s)(-Xmx\\d+[kmgtKMGT]?)\\b", jvm_config)
+        if not candidates:
+            return 0
+        return self._parse_human_readable_units(candidates[-1][-1].removeprefix("-Xmx"))
```

Comment:
I used to do that in the WordPress charm and the result is not very pleasant, you can check the `_check_addon_type` in WordPress charm.
Since this is a very short function and it only repeats twice, I think we can keep them for now. If the function expanded or more similar functions are added, we can refactor this then.

Reply:
No sure I get it, you could make a method:
```
def _extract_jvm_memory_flag(self, jvm_config: str, flag: str) -> int:
   """somedocstrings"""
   candidates = re.findall(f"(^|\\s)(-{flag}\\d+[kmgtKMGT]?)\\b", jvm_config)
   if not candidates:
      return 0
   return self._parse_human_readable_units(candidates[-1][-1].removeprefix(f"-{flag}"))
```

And then call it
`self._extract_jvm_memory_flag(jvm_config, "Xmx")` and `self._extract_jvm_memory_flag(jvm_config, "Xms")`

Reply:
Well, this isn't exactly a flag, so the naming could be a little confusing, and it probably still needs some protective programming checks to make sure the `flag` argument is `"Xmx"` or `"Xms"`.

Reply:
It's a private method the risk is limited...
At least could be put the regex in a constant ?

Reply:
I found the perfect solution for this, see the update commit.